cloudinstall/ev.py

```diff
@@ -0,0 +1,112 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import urwid
+import sys
+import cloudinstall.utils as utils
+from cloudinstall.state import ControllerState
+
+
+class EventLoop:
+
+    """ Abstracts out event loops in different scenarios
+
+    TODO: finish asyncio implementation when urwid 1.3.0
+    becomes available.
+    """
+
+    def __init__(self, ui, opts, config, log):
+        self.ui = ui
+        self.opts = opts
+        self.config = config
+        self.loop = self._build_loop()
+        self.log = log
+        self.error_code = 0
+
+    def _build_loop(self):
+        """ Returns event loop depending on output stream """
+        if self.config.getopt('headless'):
+            loop = None
+        else:
+            loop = urwid.MainLoop(self.ui, self.config.STYLES,
+                                  unhandled_input=self.header_hotkeys)
+            utils.make_screen_hicolor(loop.screen)
+            loop.screen.register_palette(self.config.STYLES)
+        return loop
+
+    def header_hotkeys(self, key):
+        if not self.config.getopt('headless'):
+            if key in ['j', 'down']:
+                self.ui.focus_next()
+            if key in ['k', 'up']:
+                self.ui.focus_previous()
+            if key == 'esc':
+                self.ui.hide_widget_on_top()
+            if key in ['h', 'H', '?']:
+                self.ui.show_help_info()
+            if key in ['a', 'A', 'f6']:
+                if self.config.getopt('controller_state') != \
+                   ControllerState.SERVICES:
+                    return
+                charm_modules = utils.load_charms()
+                charm_classes = [m.__charm_class__ for m in charm_modules
+                                 if m.__charm_class__.allow_multi_units and
+                                 not m.__charm_class__.disabled]
+                self.ui.show_add_charm_info(charm_classes, self.add_charm)
+            if key in ['q', 'Q']:
+                self.exit(0)
+            if key in ['r', 'R', 'f5']:
+                self.ui.status_info_message("View was refreshed")
+                self.update()
+
+    def exit(self, err=0):
+        self.error_code = err
+        self.log.info("Stopping eventloop")
+        if self.config.getopt('headless'):
+            sys.exit(err)
+        else:
+            urwid.ExitMainLoop()
```

Comment:
ExitMainLoop() is just an exception, this needs to be `raise urwid.ExitMainLoop()`